### ISSUE TITLE
Amr::regrid() vs AmrCore::regrid()

### ISSUE THREAD

**Issue by nmnobre** [AMReX-Codes/amrex#3014]
Hi,

In what circumstances should we be building code that uses each of these classes/functions in our multi-level AMR code?

(I know I have an unrelated open issue which I still didn't have the time to reply to, but I do appreciate the help I was given there, I'll get to it. 🙂)

Cheers,
-Nuno

**Comment by WeiqunZhang** [comment]
https://amrex-codes.github.io/amrex/docs_html/Faq.html

> Q. What’s the difference between AmrCore and AmrLevel? How do I decide which to use?
>
 > A. The AmrLevel class is an abstract base class that holds data for a single AMR level. A vector of AmrLevel is stored in the Amr class, which is derived from AmrCore. An application code can derive from AmrLevel and override functions. AmrCore contains the meta-data for the AMR hierarchy, but it does not contain any floating point mesh data. Instead of using Amr/AmrLevel, an application can also derive from AmrCore. If you want flexibility, you might choose the AmrCore approach, otherwise the AmrLevel approach might be easier because it already has a lot of built-in capabilities that are common for AMR applications.

Does this answer your question?

**Comment by nmnobre** [comment]
Partially, yes, sorry I should have probably said I was aware of that FAQ.

Let me rephrase, the functionality in both `regrid()` functions doesn't seem exactly the same, in particular with regards to the load balancing with work estimates?

**Comment by WeiqunZhang** [comment]
If your code is not using `Amr`, you will not be able to use `Amr::regrid`.  So I assume you have a code that uses `Amr`.  In that case, it's possible to use `AmrCore::regrid`, but you need to have a class that derives from `Amr` and implement a number of virtual functions.  Given that, I think you should just use `Amr::regird`.

As for load balancing, `AmrCore` does not have access to any work estimates collected by the application code, so it cannot do any load balance based on that.  However, that does NOT mean `AmrCore` cannot do any load balance.  In the `RemakeLevel` virtual function, the user is allowed to ignore the provided `DistributionMapping` and uses their own `DistributionMapping`.  They just need to remember to call `SetDistributionMapping` to let `AmrCore` know about it.

For `Amr`, you already know that there is an option to make load balanced DistributionMapping based on work estimate.  If that does not work as you like, you can ignore the DistributionMapping provided to your LevelBld, and use what you like when building a new AmrLevel.  You don't have to let Amr/AmrLevel know, because they already know about it.

**Comment by nmnobre** [comment]
Thanks, this is actually very useful.

> As for load balancing, `AmrCore` does not have access to any work estimates collected by the application code, so it cannot do any load balance based on that. However, that does NOT mean `AmrCore` cannot do any load balance. In the `RemakeLevel` virtual function, the user is allowed to ignore the provided `DistributionMapping` and uses their own `DistributionMapping`. They just need to remember to call `SetDistributionMapping` to let `AmrCore` know about it.

1) Will this waste the DM computation time AMReX must do, say for SFC? This should be minor, but still...
2) As I understand it, the work estimate will be some `StateData` and this could potentially be any data I choose to store over my boxes, even data with physical meaning. In your experience, what would you use in a practical application?

Thank you.

**Comment by WeiqunZhang** [comment]
1.  That is true.
2. The work estimate StateData based load balance in `Amr` class was added when we were developing EB (embedded boundary).  We used the wall time to set the work estimate.  For EB, some boxes may have zero work because their cells are all covered.  Many boxes have a mixture of regular, cut and covered cells.  It's hard to estimate the cost base on the number of cells.  Using the wall time seemed to work well.  The downside is one has to call `Gpu::streamSynchronize()` to actually get the true cost of GPU kernels.  See amrex/Tests/EB_CNS/Source/CNS_advance.cpp for an example of how the cost is estimated.

**Comment by WeiqunZhang** [comment]
> Will this waste the DM computation time AMReX must do, say for SFC? This should be minor, but still...

We could add a virtual function to AmrCore for making DM.

**Comment by nmnobre** [comment]
> We could add a virtual function to AmrCore for making DM.

That is true!

**Comment by nmnobre** [comment]
> * The work estimate StateData based load balance in `Amr` class was added when we were developing EB (embedded boundary).  We used the wall time to set the work estimate.  For EB, some boxes may have zero work because their cells are all covered.  Many boxes have a mixture of regular, cut and covered cells.  It's hard to estimate the cost base on the number of cells.  Using the wall time seemed to work well.  The downside is one has to call `Gpu::streamSynchronize()` to actually get the true cost of GPU kernels.  See amrex/Tests/EB_CNS/Source/CNS_advance.cpp for an example of how the cost is estimated.

I see, you worked around the fact that you actually only have a weight per box by computing an average over the entire box and assigning it to every single one of its elements. So that when you do `gather_weights` you recover the box weight. Perfect.
